flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -360,32 +360,44 @@ public boolean notifyBufferAvailable(Buffer buffer) {
 			return false;
 		}
 
-		boolean needMoreBuffers = false;
-		synchronized (bufferQueue) {
-			checkState(isWaitingForFloatingBuffers, "This channel should be waiting for floating buffers.");
+		boolean recycleBuffer = true;
+		try {
+			boolean needMoreBuffers = false;
+			synchronized (bufferQueue) {
+				checkState(isWaitingForFloatingBuffers,
+					"This channel should be waiting for floating buffers.");
+
+				// Important: double check the isReleased state inside synchronized block, so there is no
+				// race condition when notifyBufferAvailable and releaseAllResources running in parallel.
+				if (isReleased.get() ||
+					bufferQueue.getAvailableBufferSize() >= numRequiredBuffers) {
+					isWaitingForFloatingBuffers = false;
+					buffer.recycleBuffer();
+					return false;
+				}
 
-			// Important: double check the isReleased state inside synchronized block, so there is no
-			// race condition when notifyBufferAvailable and releaseAllResources running in parallel.
-			if (isReleased.get() || bufferQueue.getAvailableBufferSize() >= numRequiredBuffers) {
-				isWaitingForFloatingBuffers = false;
-				buffer.recycleBuffer();
-				return false;
-			}
+				recycleBuffer = false;
```

Comment:
I don't understand this `recycleBuffer` flag. How can it ever be `true`?
- If we throw form `checkState` - that case we should just ignore. 
- If we throw from `bufferQueue.getAvailableBufferSize()`? Is it possible?
- If we throw from `buffer.recycleBuffer();`? In that case for whatever reason we attempt to recycle second time?

none of those options make sense, so am I missing something?

Reply:
I don't think we should ignore the `checkState` since a failure there in our task is of course an implementation error but should not leak buffers for other tasks still running and working with this buffer pool.
Regarding the `buffer.recycleBuffer()`, you are right: we kind of always assume this cannot throw, but here we should maybe guard against recycling twice.